File: GFSSOC Python/gfssoc1j5_pursuit_of_knowledge/GFSSOC1J5.py
import sys
from collections import deque
input = sys.stdin.readline


class Digraph:
    def __init__(self, V):
        self.__V = V
        self.__E = 0
        self.__adjList = []
        for v in range(V):
            self.__adjList.append([])
        self.__indegreeArr = [0] * V

    def getV(self):
        return self.__V

    def getE(self):
        return self.__E

    def addEdge(self, v, w):
        self.__adjList[v].append(w)
        self.__indegreeArr[w] += 1
        self.__E += 1

    def adj(self, v):
        return self.__adjList[v]

    def outdegree(self, v):
        return len(self.__adjList[v])

    def indegree(self, v):
        return self.__indegree[v]


class BreadthFirstSearch:
    def __init__(self, G, s):
        self.__marked = [False] * G.getV()
        self.__distTo = [float('inf')] * G.getV()
        self.__bfs(G, s)

    def __bfs(self, G, s):
        q = deque()
        self.__distTo[s] = 0
        self.__marked[s] = True
        q.append(s)
        while len(q) != 0:
            v = q.popleft()
            for w in G.adj(v):
                if not self.__marked[w]:
                    self.__distTo[w] = self.__distTo[v] + T
                    self.__marked[w] = True
                    q.append(w)

    def getDistTo(self, v):
        return self.__distTo[v]


N, M, T = map(int, input().split())
G = Digraph(N)
bfs = [None] * N
for i in range(M):
    a, b = map(int, input().split())
    a -= 1
    b -= 1
    G.addEdge(a, b)
for i in range(N):
    bfs[i] = BreadthFirstSearch(G, i)
Q = int(input())
for i in range(Q):
    a, b = map(int, input().split())
    a -= 1
    b -= 1
    ans = bfs[a].getDistTo(b)
    if ans == float('inf'):
        print('Not enough hallways!')
    else:
        print(ans)

File: GFSSOC C++/gfssoc1j5_pursuit_of_knowledge/GFSSOC1J5.cpp
/*
 * GFSSOC1J5.cpp
 *
 *  Created on: Jul 3, 2017
 *      Author: Wesley Leung
 */

#include <bits/stdc++.h>

#define SHORT_INF 0x3f3f
#define INT_INF 0x3f3f3f3f
#define LL_INF 0x3f3f3f3f3f3f3f3f
#define pb push_back
#define mp make_pair
#define l(x) x << 1
#define r(x) x << 1 | 1
#define mid(x, y) x + (y - x) / 2
#define f first
#define s second
#define ri(x) scanf("%d", &x)
#define rll(x) scanf("%lld", &x)
#define rllu(x) scanf("%llu", &x)
#define rf(x) scanf("%f", &x)
#define rd(x) scanf("%lf", &x)
#define rc(x) scanf("%c", &x)
#define rs(x) scanf("%s", x)
#define Fill(a, x) memset(a, x, sizeof(a))
#define randi(a, b) rand() % (b - a + 1) + a

using namespace std;

typedef long long ll;
typedef unsigned long long llu;
typedef pair<int, int> pii;
typedef pair<float, float> pff;
typedef pair<double, double> pdd;
typedef pair<ll, ll> pll;
typedef pair<llu, llu> pllu;
typedef map<int, int> mii;
typedef map<int, ll> mill;
typedef map<ll, int> mlli;
typedef unordered_map<int, int> umii;
typedef unordered_map<int, ll> umill;
typedef unordered_map<ll, int> umlli;

struct Digraph {
private:
    int V;
    int E;
    vector<int>* adjList;
    int *indegreeArr;

public:
    Digraph(int V) {
        this->V = V;
        this->E = 0;
        adjList = new vector<int>[V];
        indegreeArr = new int[V];
    }

    int getV() {
        return this->V;
    }

    int getE() {
        return this->E;
    }

    void addEdge(int v, int w) {
        adjList[v].push_back(w);
        indegreeArr[w]++;
        E++;
    }

    vector<int> &adj(int v) {
        return adjList[v];
    }

    int outdegree(int v) {
        return adjList[v].size();
    }

    int indegree(int v) {
        return indegreeArr[v];
    }
};

#define MAXN 1000

Digraph *G;
bool marked[MAXN];
int N, M, T, Q, a, b, distTo[MAXN][MAXN];

void bfs(int s) {
    queue<int> q;
    for (int v = 0; v < G->getV(); v++) {
        marked[v] = false;
        distTo[s][v] = INT_MAX;
    }
    distTo[s][s] = 0;
    marked[s] = true;
    q.push(s);

    while(!q.empty()) {
        int v = q.front();
        q.pop();
        for (int w : G->adj(v)) {
            if (marked[w]) continue;
            distTo[s][w] = distTo[s][v] + T;
            marked[w] = true;
            q.push(w);
        }
    }
}

int main() {
    ri(N);
    ri(M);
    ri(T);
    G = new Digraph(N);
    for (int i = 0; i < M; i++) {
        ri(a);
        ri(b);
        a--;
        b--;
        G->addEdge(a, b);
    }
    for (int i = 0; i < N; i++) {
        bfs(i);
    }
    ri(Q);
    for (int i = 0; i < Q; i++) {
        ri(a);
        ri(b);
        a--;
        b--;
        int ans = distTo[a][b];
        if (ans == INT_MAX) printf("Not enough hallways!\n");
        else printf("%d\n", distTo[a][b]);
    }
    return 0;
}
